core/src/test/java/com/linecorp/armeria/server/HttpServerKeepAliveHandlerTest.java

```diff
@@ -0,0 +1,113 @@
+import java.util.concurrent.atomic.AtomicInteger;
+
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.extension.RegisterExtension;
+import org.junit.jupiter.params.ParameterizedTest;
+import org.junit.jupiter.params.provider.CsvSource;
+
+import com.linecorp.armeria.client.ClientFactory;
+import com.linecorp.armeria.client.ConnectionPoolListener;
+import com.linecorp.armeria.client.WebClient;
+import com.linecorp.armeria.common.HttpResponse;
+import com.linecorp.armeria.common.SessionProtocol;
+import com.linecorp.armeria.server.logging.LoggingService;
+import com.linecorp.armeria.testing.junit.server.ServerExtension;
+
+import io.netty.util.AttributeMap;
+
+class HttpServerKeepAliveHandlerTest {
+
+    private static final int serverIdleTimeout = 5000;
+    private static final int serverPingInterval = 1000;
+
+    @RegisterExtension
+    static ServerExtension server = new ServerExtension() {
+        @Override
+        protected void configure(ServerBuilder sb) throws Exception {
+            sb.idleTimeoutMillis(serverIdleTimeout);
+            sb.pingIntervalMillis(serverPingInterval);
+            sb.decorator(LoggingService.newDecorator())
+              .service("/", (ctx, req) -> HttpResponse.of("OK"));
+        }
+    };
+
+    private AtomicInteger counter;
+    private ConnectionPoolListener listener;
+
+    @BeforeEach
+    void setUp() {
+        counter = new AtomicInteger();
+        listener = new ConnectionPoolListener() {
+            @Override
+            public void connectionOpen(SessionProtocol protocol, InetSocketAddress remoteAddr,
+                                       InetSocketAddress localAddr, AttributeMap attrs) throws Exception {
+                counter.incrementAndGet();
+            }
+
+            @Override
+            public void connectionClosed(SessionProtocol protocol, InetSocketAddress remoteAddr,
+                                         InetSocketAddress localAddr, AttributeMap attrs) throws Exception {
+                counter.decrementAndGet();
+            }
+        };
+    }
+
+    @ParameterizedTest
+    @CsvSource({ "H1C", "H2C" })
+    void closeByClientIdleTimeout(SessionProtocol protocol) throws InterruptedException {
+        final int clientIdleTimeout = 2000;
+        final ClientFactory factory = ClientFactory.builder()
+                                                   .idleTimeoutMillis(clientIdleTimeout)
+                                                   .connectionPoolListener(listener)
+                                                   .build();
+        final WebClient client = WebClient.builder(server.uri(protocol))
+                                          .factory(factory)
+                                          .build();
+
+        client.get("/").aggregate().join();
+
+        // HTTP/2 PING frame sent by the server cannot prevent to close an idle connection.
+        Thread.sleep(clientIdleTimeout + 1000);
```

Comment:
I think this test will make flaky test. How about using just `await()` after `clientIdleTimeout` elapsed?

Reply:
Unfortunately, we should not use `await()` because `clientIdleTimeout` is `2000` and `serverIdleTimeout` is `5000`.
I wanted to check whether the connection is closed by the client-side.
If it waits more than 5000 milliseconds here, we don't know which closes the connection.
Probably we could multiply the timeouts by 2 and use `await().timeout(serverIdleTimeout - buffer)`.

Reply:
Then How about just checking if `serverIdleTimeout` is passed after the counter becomes `0`?
Also, don't we need to check if the counter was `1` before it gets to `0`?

Reply:
Oh, that sounds good.